src/task.c

```diff
@@ -133,23 +168,86 @@ task_register_hook(task_hook newhook)
    return STATUS_OK;
 }
 
-// Register the calling thread and allocate scratch space for it
+/*
+ * task_register_thread(): Register a new thread with the task system.
+ *
+ * Registration implies:
+ *  - Acquire a new thread ID (index) for this to-be-active thread
+ *  - Allocate scratch space for this thread, and track this space..
+ */
 void
-task_register_this_thread(task_system *ts, uint64 scratch_size)
+task_register_thread(task_system *ts,
+                     uint64       scratch_size,
+                     const char  *file,
+                     const int    lineno,
+                     const char  *func)
 {
+   threadid thread_tid = platform_get_tid();
+   /*
+    * For all newly created threads, the thread-index will be 0 before the hook
+    * function, run below, generates this thread's index. We do not want to
+    * register a thread twice; otherwise, it will simply burn up an index and
+    * possibly allocated scratch space.
+    * As threads generally will require scratch space to be allocated, an
+    * attempt to re-register a thread should find scratch space already
+    * allocated. Flag that.
+    */
+   platform_assert((thread_tid == 0)
+                      || (ts->thread_scratch[thread_tid] == NULL),
+                   "[%s:%d::%s()] Thread at index %lu is found to have scratch"
+                   " space already allocated, %p\n",
+                   file,
+                   lineno,
+                   func,
+                   thread_tid,
+                   ts->thread_scratch[thread_tid]);
+
    char *scratch = NULL;
    if (scratch_size > 0) {
       scratch = TYPED_ZALLOC_MANUAL(ts->heap_id, scratch, scratch_size);
    }
    task_run_thread_hooks(ts);
-   ts->thread_scratch[platform_get_tid()] = scratch;
+
+   // Ensure that we are not clobbering some scratch space previously allocated
+   // that might occur due to a code-bug or by an incorrectly generated tid.
+   thread_tid = platform_get_tid();
+   platform_assert(
+      (ts->thread_scratch[thread_tid] == NULL),
+      "[%s:%d::%s()] Scratch space found allocated at %p for thread with "
+      "index %lu.",
+      file,
+      lineno,
+      func,
+      ts->thread_scratch[thread_tid],
+      thread_tid);
+
+   ts->thread_scratch[thread_tid] = scratch;
 }
 
-// Deregister the calling thread and free its scratch space
+/*
+ * task_deregister_thread() - Deregister an active thread from the task system.
+ *
+ * Deregistration involves:
+ *  - Releasing any scratch space acquired for this thread.
+ *  - Clearing the thread ID (index) for this thread
+ */
 void
-task_deregister_this_thread(task_system *ts)
+task_deregister_thread(task_system *ts,
+                       const char  *file,
+                       const int    lineno,
+                       const char  *func)
 {
-   uint64 tid = platform_get_tid();
+   threadid tid = platform_get_tid();
+
+   // A registered thread should always have a non-zero tid (index).
+   if (tid == 0) {
```

Comment:
Useful check ... Found bugs in test code that was doing 2 de-registers, inadvertently.

Reply:
Good catch